### DISCUSSION TITLE
Python REPL

### DISCUSSION THREAD

**Discussion by gswifort** [CEXT-Dan/PyRx#91] [Ideas]
Would it be possible to run something similar to `PYCMDPROMPT` (REPL) in an external console, preferably in the vscode console (`integratedConsole`).  It's a bit like running `debugpy`, but without locking the AutoCAD interface.

**Comment by CEXT-Dan** [comment]
I don’t know, that would be cool though

Debugpy, is all Python, or at least I don’t do anything in C++ to enable this except redirect stdout
https://github.com/CEXT-Dan/PyRx/blob/d7a9404c89569edc3003dcc4321d70c4fc95a8bc/PyRxCore/PyAcRx.cpp#L36-L89

Communication with Debugpy is not without its side effects. I’m pretty sure that Debugpy hijacks stdin to push data to python. But if you notice when debugging, interaction with AutoCAD’s command line gets wonky 

Ideally, I would like to see something that can be more interactive with the CAD side like autolisp
I would also like to see this done in Python. The goal is to expose AutoCAD to Python so these sorts of tasks could be done in Python 

A short-term solution would be to create a modeless dialog, or even better a palette (Like the PyChat sample, maybe with the send button removed and have an ‘enter event’)

**Comment by gswifort** [comment]
So, let's do it!
Below is a demo.

1. Load the module into CAD:

    ```py
    import code
    import sys
    import threading
    
    import win32console
    
    from pyrx_imp import *
    
    
    def PyRxCmd_interact():
    
        threading.Thread(target=main).start()
    
    
    def main():
        try:
            win32console.AllocConsole()
            f_in = open("CONIN$", "r")
            f_out = open("CONOUT$", "w")
            sys.stdout = f_out
            sys.stderr = f_out
            sys.stdin = f_in
    
            code.interact(local=globals())
    
        finally:
            win32console.FreeConsole()
    ```

2. Call `INTERACT`. A console window should open.

3. Do anything in the console, e.g.:

    ```py
    >>> p = (
        Ap.Application()
        .docManager()
        .curDocument()
        .editor()
        .getPoint("\nGetPoint\n")
    )
    >>> print(p)
    ```

    and now the strange thing... at the beginning it worked for me, the console waited for me to click on the point and returned it. Then it stopped working - the command is sent to CAD, the point cursor appears, but the console immediately raises `RuntimeError: unidentifiable C++ exception`.
    Maybe I should use this part somehow: https://github.com/CEXT-Dan/PyRx/blob/4583ba3233f40f84704668d8d3dbecc5b4bfeee9/PySamples/wxWidgets/EditorInteraction/testwxpyxrcmodeless.py#L61-L71

4. When you finish your actions in the console, type `exit()` to release the console, otherwise when you close the console, CAD will also close.


Additionally, I don't know how to add something to the model space of the current document, I get a message that the database is locked.

**Reply by gswifort** [reply]
Ok, now I know why it worked for me the first time. I loaded [your module](https://github.com/CEXT-Dan/PyRx/blob/main/PySamples/wxWidgets/EditorInteraction/testwxpyxrcmodeless.py), ran the command `WXPYXRCM`, closed the dialog, loaded my module, ran `INTERACT` , in the console I ran:
```py
>>> p = (
    Ap.Application()
    .docManager()
    .curDocument()
    .editor()
    .getPoint("\nGetPoint\n")
)
>>> print(p)
```
With this combination, the console waits for the CAD to indicate a point.

**Comment by gswifort** [comment]
This works very well:

```py
import traceback

import wx
import wx.py.shell
from pyrx_imp import Ap, Db, Ed, Ge, Gi, Gs, Rx  # will be added to PyShell


def PyRxCmd_pyshell():
    try:
        dlg = wx.py.shell.ShellFrame(None, locals=globals())
        dlg.Show()
    except Exception:
        traceback.print_exc()
```

![image](https://github.com/user-attachments/assets/bf9b757a-0405-4f83-a078-cd272531b4f0)

-----------------------------
I still don't know how to add something to the database:
```py
>>> p1 = Ge.Point3d(0,0,0)
>>> p2 = Ge.Point3d(100,100,0)
>>> line = Db.Line(p1,p2)
>>> Db.curDb().addToCurrentspace(line)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
RuntimeError: 
Exception!(eLockViolation), function addToBlock1, Line 560, File PyDbDatabase.cpp: 
```

**Reply by CEXT-Dan** [reply]
That's so cool! 
for the lock just add Ap.AutoDocLock()
I don't know when garbage collection happens though

**Reply by CEXT-Dan** [reply]
Excellent find! This works very well indeed.
I’ll take a look at 3,4,5 

1, It has IntelliSense, but it doesn’t seem to read the stub files
2. It needs to be locked to the current document? Maybe not?  
3, it needs to ‘switch to’ the cad application on interactions like entget. 
4, scope needs to be investigated 
5. Investigate if this can be moved into a palette or a control bar, this will take care of the ‘switch to’ and make the frame dock-able to Acad’s main frame

![pyshell](https://github.com/user-attachments/assets/875192f5-a668-4bd7-97f7-47a59dad9906)

**Reply by gswifort** [reply]
> for the lock just add Ap.AutoDocLock()

It works. The line is added but is not visible on the screen. It only appears after saving and reopening the document.

**Reply by CEXT-Dan** [reply]
In AutoCAD? 
For me, it shows once I click on the drawing. 

But It’s undefined behavior in all platforms, because it’s in a global scope and never gets unlocked 

So, maybe 
lock = Ap.AutoDocLock()   
…
Del(lock)

Palettes are also modeless, gracefully handle WM_ACAD_KEEPFOCUS, otherwise we might have to add it to PyShell. Allow AutoCAD to become active when the mouse leaves the dialog 
https://github.com/CEXT-Dan/PyRx/blob/4583ba3233f40f84704668d8d3dbecc5b4bfeee9/PySamples/wxWidgets/EditorInteraction/testwxpyxrcmodeless.py#L61-L78

**Reply by CEXT-Dan** [reply]
try this for zwcad
```Python
lock = Ap.AutoDocLock()
line = Db.Line(Ge.Point3d(0,0,0),Ge.Point3d(100,100,0))
Db.curDb().addToModelspace(line)
del(line)
del(lock)
```

**Reply by CEXT-Dan** [reply]
So GC will be an issue for database objects.  User will need to manually call DbObject.close(), or dispose()
use the ActiveX wrappers, or add a scope
```Python
def foo():
    lock = Ap.AutoDocLock()
    line = Db.Line(Ge.Point3d(0,0,0),Ge.Point3d(100,100,0))
    Db.curDb().addToModelspace(line)
```
![pyshell](https://github.com/user-attachments/assets/6188152c-0fdc-46d4-9a51-9e47f1bc32db)

**Reply by gswifort** [reply]
It works! Thanks!

**Reply by gswifort** [reply]
> User will need to manually call DbObject.close(), or dispose() use the ActiveX wrappers, or add a scope

Or this:

```py
import inspect
import traceback

import wx
import wx.py.shell
from pyrx_imp import Ap, Db, Ed, Ge, Gi, Gs, Rx  # will be added to PyShell


class DbTransaction:
    def __enter__(self):
        self.frame = inspect.currentframe().f_back
        self.enter_vars = set(self.frame.f_locals.keys())
        self.lock = Ap.AutoDocLock()
        return Db.curDb()

    def __exit__(self, exc_type, exc_val, exc_tb):
        exit_vars = set(self.frame.f_locals.keys())
        new_vars = exit_vars - self.enter_vars
        for var in new_vars:
            del self.frame.f_locals[var]
        del self.lock


def PyRxCmd_pyshell():
    try:
        dlg = wx.py.shell.ShellFrame(None, locals=globals())
        dlg.Show()
    except Exception:
        traceback.print_exc()
```

in PyShell:

```py
with DbTransaction() as db:
    line = Db.Line(Ge.Point3d(0,0,0),Ge.Point3d(100,100,0))
    db.addToModelspace(line)
```

**Reply by CEXT-Dan** [reply]
Super cool!

**Comment by CEXT-Dan** [comment]
PyShell, calling help(), copyright(), credits(), license(), nukes AutoCAD

**Reply by CEXT-Dan** [reply]
seems these can be overridden 
```Python
def help():
    print("No help for you")
    
def copyright():
    print("No copyright for you")
    
def credits():
    print("No credits for you")
    
def license():
    print("No license for you")

def PyRxCmd_pyshell():
...
```

**Comment by CEXT-Dan** [comment]
I tried to improve PyEdUserInteraction::userInteraction,for function like entSel(“\nSelect: “)
 but failed, it turned ShellFrame modal 

ShellFrame inherits from a custom frame, that may be the place to start customizing
I.e. on mouse leave, set focus to parent ( AutoCAD)

**Reply by gswifort** [reply]
I guess I don't understand what exactly you are trying to achieve.

**Reply by CEXT-Dan** [reply]
Not a big deal really, but to illustrate the behavior, see testwxpyxrc
PySamples\wxWidgets\EditorInteraction\testwxpyxrc.py
 
Is a modal dialog with getpoint, and getdist
userInteraction class automatically hides child dialogs, sets focus to AutoCAD’s editor,
then restores the dialog hierarchy. 

Since userInteraction is already embedded in all the editor functions, I thought that might be the place 
So set focus from PyShell to AutoCAD, when editor commands are run.

**Comment by gswifort** [comment]
PyShell on the palette:

```py
import traceback

import wx
from pyrx_imp import *
from wx.py.shell import Shell

palette = Ap.PaletteSet("PyShell")


class MyPanel(wx.Panel):
    def __init__(self):
        super().__init__()
        self.Bind(wx.EVT_SHOW, self.OnShow)

    def OnShow(self, event):
        sizer = wx.BoxSizer()
        sizer.Add(Shell(self, locals=globals()), 1, wx.EXPAND, 0)
        self.SetSizerAndFit(sizer)
        self.Layout()


def createPalette():
    try:
        panel = MyPanel()
        palette.add("PyShell", panel)
        palette.setVisible(True)
        palette.setName("PyShell")

    except Exception:
        traceback.print_exc()


def PyRxCmd_pyshellpalette():
    try:
        createPalette()
    except Exception:
        traceback.print_exc()
```

**Reply by CEXT-Dan** [reply]
Yes! That’s amazing!
BTY, in AutoCAD, BricsCAD, you can add a GUID, to persist the location and dock state

```Python
palette = Ap.PaletteSet("PyShell","{b98f3a34-ed66-4c3f-9c6a-9804e13b18b0}")
```

**Reply by CEXT-Dan** [reply]
BTY, I improved palette docking
Added palette.enableDocking(Ap.PaletteDockStyle.kANY), kANY is default 
Next release, should be able to dock just like the command line

**Reply by gswifort** [reply]
> BTY, in AutoCAD, BricsCAD, you can add a GUID, to persist the location and dock state

There is no way to persist the location of the pallet in ZWCAD?

**Reply by CEXT-Dan** [reply]
Next release, you should be able to code the docking side and size, having a bit of trouble undocking  

```Python
def createPalette():
    try:
        panel = MyPanel()
        palette.add("PyShell", panel)
        palette.setVisible(True)
        
        host = Ap.Application.hostAPI()
        if 'ZRX' in host or 'GRX' in host:
            palette.setSize(100,100)
            palette.dock(Ap.PaletteDockStyle.kBottom)
        
    except Exception:
        traceback.print_exc()
```

![zwshell](https://github.com/user-attachments/assets/8c1a968c-b126-43a7-870b-7c3c9e9d6641)

**Reply by gswifort** [reply]
Is it possible to save the location set by the user, preferably with an onMove event?  Even in a separate file or in XData.

**Reply by CEXT-Dan** [reply]
need to file a bug report with ZwCAD, that it does not honor SetToolID in RestoreControlBar

One approach I would try, is to use is store the data in setEnv. I’d probably need to do more work on palettes get the dock state
but something like

```Python
host = Ap.Application.hostAPI()
palette = Ap.PaletteSet("PyShell","{b98f3a34-ed66-4c3f-9c6a-9804e13b18b0}")

paletteData = str
def OnPyInitApp():
    global paletteData
    if 'ZRX' in host or 'GRX' in host:
        paletteData = Ed.Core.getEnv("PYSHELLPALETTE")
    
def OnPyUnloadApp():
    if 'ZRX' in host or 'GRX' in host:
            palette.getDockState...
            Ed.Core.setEnv("PYSHELLPALETTE", paletteData)
```

**Reply by gswifort** [reply]
Can you capture the palette closing event (X click)?  When opening the palette again in the same session, the location is also not persist.

**Reply by CEXT-Dan** [reply]
I’ll need to investigate this, I had only prototyped palettes as a proof of concept.
.NET has separate event handlers, I would like to see if I can forward events though the message pump 
I.e.  OnShow,  OnClose

**Reply by CEXT-Dan** [reply]
Unfortunately, I have not found a good way to fire the event for the palette close button. 
It’s marked as internal in ARX, neither WM_CLOSE nor SC_CLOSE worked. I found the I could hook into the messages and get the notification, But the message ID is different for every platform, so I just reverted it. This could be done in python wndproc 

AutoCAD only persists the position per session, so if you move, close, restore, it’s back to the previous position. 
I’ve added a few new functions though 

```Python
def PyRxCmd_doit():
    try:
        print(palette.getDockState()) #new
        print(palette.getFullRect())
        
        frame : wx.Frame = palette.getWxFrame() #new 
        print(frame.Size)
        print(frame.Name)
        
    except Exception:
        traceback.print_exc()
        
# kBottom
# (0, 959, 2560, 1229)
# (2560, 270)
# PyShell

```

**Reply by gswifort** [reply]
I understand that we are still waiting for `setDockState` to be fully implemented?

**Reply by CEXT-Dan** [reply]
Yeah, dock state none, or set to floating, also setLocation has issues after the palette is shown. It should work from a restore perspective.

**Reply by CEXT-Dan** [reply]
I have setLocation and setDockState implemented, 
I also noticed I could have introduced a bug where OnSize may have covered part of the palette, this is fixed
Difficult to get right, every platform has it’s subtle differences

**Reply by gswifort** [reply]
This works quite well (for persist location) on ZWCAD. It will probably work even better with the next release where `setLocation` will be improved.

```py
import pickle
import traceback
from pathlib import Path

import win32api
import win32con
import win32gui
import wx
from pyrx_imp import *
from wx.py.shell import Shell

BASE_DIR = Path(__file__).parent


class MyPaletteSet(Ap.PaletteSet):
    POSITION_FILE = BASE_DIR / "pyshell_palette.pickle"

    def __init__(
        self,
        name: str = "PyShell",
        guid: str = "{29d65d38-1c5f-4a49-8ff9-7a37fcfab3d9}",
    ) -> None:
        super().__init__(name, guid)
        self._wm_capture_changed = False

    def addHook(self):
        self.oldWndProc = win32gui.SetWindowLong(
            self.getWxWindow().GetHandle(), win32con.GWL_WNDPROC, self.MyWndProc
        )

    def restoreHook(self):
        return win32api.SetWindowLong(
            self.getWxWindow().GetHandle(), win32con.GWL_WNDPROC, self.oldWndProc
        )

    def OnClose(self, event):
        self.restoreHook()

    def MyWndProc(self, hWnd, msg, wParam, lParam):
        if msg == win32con.WM_CAPTURECHANGED:
            # for PaletteDockStyle.kNone
            self.save_position()
            self._wm_capture_changed = True
        elif self._wm_capture_changed and msg == win32con.SW_SHOWMAXIMIZED:
            # for other PaletteDockStyle
            self.save_position()
            self._wm_capture_changed = False

        return win32gui.CallWindowProc(self.oldWndProc, hWnd, msg, wParam, lParam)

    def save_position(self):
        window: wx.Window = self.getWxWindow()
        data = {
            "size": window.Size,
            "position": (*window.Position,),
            "dock": self.getDockState(),
        }
        with open(self.POSITION_FILE, "wb") as f:
            pickle.dump(data, file=f)

    def load_position(self):
        if not self.POSITION_FILE.exists():
            return
        with open(self.POSITION_FILE, "rb") as f:
            data = pickle.load(file=f)
        window: wx.Window = self.getWxWindow()
        if (dock := data["dock"]) == Ap.PaletteDockStyle.kNone:  # NotImplementedYet
            return
        self.setDockState(dock)
        window.SetSize(data["size"])
        self.setLocation(*data["position"])


class MyPanel(wx.Panel):
    def __init__(self):
        super().__init__()
        self.Bind(wx.EVT_SHOW, self.OnShow)

    def OnShow(self, event):
        sizer = wx.BoxSizer()
        sizer.Add(Shell(self, locals=globals()), 1, wx.EXPAND, 0)
        self.SetSizerAndFit(sizer)
        self.Layout()


palette = None
pyshellpanel = None
pyshellpanel_id = None


def addPyShellPanel(palette: MyPaletteSet):
    global pyshellpanel
    global pyshellpanel_id
    if pyshellpanel is not None:
        pyshellpanel.Destroy()
    if pyshellpanel_id is not None:
        palette.removePalette(pyshellpanel_id)
    pyshellpanel = MyPanel()
    pyshellpanel_id = palette.add("PyShell", pyshellpanel)
    palette.setName("PyShell")


def PyRxCmd_pyshellpalette():
    try:
        global palette
        global_palette = palette
        if global_palette is None:
            palette = MyPaletteSet()
            addPyShellPanel(palette)
        palette.setVisible(True)
        try:
            palette.load_position()
        except Exception as e:
            print(e)
        if global_palette is None:
            palette.addHook()

    except Exception:
        traceback.print_exc()
```

**Reply by gswifort** [reply]
Test this module in your free time. Top, right and left docking works. If I set it up without docking, sometimes it works, sometimes it doesn't. Bottom docking always positions itself below the command line, even if I dock above it.
I added some logs, you can enable them by running `LOGLVL 10` in CAD.

[pyshell_palette.zip](https://github.com/user-attachments/files/16302521/pyshell_palette.zip)

_Tested on ZWCAD_

**Reply by CEXT-Dan** [reply]
I get a split of Pyshell, and the Command line in Zw2025, works great in Zw2024 and GStarCAD's
Looks like either they changed to a new docking system from Zw2024to Zw2025 and have not ironed out all the bugs

**Reply by CEXT-Dan** [reply]
BTW, this is what AutoCAD stores, persisted on application close 
```
<Tool CLSID="{B98F3A34-ED66-4C3F-9C6A-9804E13B18B0}">
	<CAdUiDockControlBar Orientation="59422" AllowDocking="1">
		<FloatInfo Left="113" Top="224" Width="1110" Height="315"/>
		<DockInfo Left="0" Top="909" Width="2560" Height="221" MRUDockID="59422"/>
	</CAdUiDockControlBar>
	<CAdUiPaletteSet ActivePaletteIndex="0" TitleBarLocation="0" AutoRollup="0" Name="PyShell" Style="4106" Opacity="48" RolloverOpacity="100" Visible="1"/>
</Tool>
```

**Reply by gswifort** [reply]
In Zwcad 2024, when it docks under the command line, it positions itself above the command line after restart. `kNone` docking sometimes loads fine and sometimes it docks to the left.

**Reply by CEXT-Dan** [reply]
There’s a function Ed.Core. getAcadDockCmdLine, this return’s the handle to the command line. Though I have not test it with zwcad 
https://help.autodesk.com/view/OARX/2023/ENU/?guid=OARX-RefGuide-acedGetAcadDockCmdLine
You maybe be able use win32api 

I’ll wrap restoreControlBar, maybe that will give you more control, additionally I will investigate wrapping acedRegisterWatchWinMsg
This will give you every message in the loop.

**Reply by CEXT-Dan** [reply]
kNone calls PostMessage(WM_COMMAND, ID_ADUI_ALLOWDOCK, 0); 0x1001 (4097)
as far as I can tell

**Reply by CEXT-Dan** [reply]
>  kNone docking sometimes loads fine and sometimes it docks to the left.

I see, I think I had a typo

**Reply by CEXT-Dan** [reply]
Thinking of also working the other direction, instead of me writing the implementation, for more generic applications, just wrapping what is needed from CAdUiDockControlBar and CControlBar, so the python developer can have finer grained control  

You can get hints of the implementation from using IlSpy 
https://github.com/icsharpcode/ILSpy

In IlSpy add ZwManaged, dill down to ZwSoft.ZwCAD.Windows.PaletteSet.Dock

**Comment by CEXT-Dan** [comment]
Autocomplete doesn't use the stubs. I asked about it here
https://discuss.wxpython.org/t/wx-py-pyshell-read-pyi/37000
there's a response, but I don't understand it

**Reply by gswifort** [reply]
Yes, I have already researched this topic as well. This will likely require a bit more work.  We will definitely have to work on generating stubs files because they are not always valid python files (I think this will be required because we will need to evaluate these files).

**Reply by CEXT-Dan** [reply]
Everything needed is in the doc string, it just needs to be parsed. If the Shell has access to the modules 
I.e. locals=globals(), then maybe it’s possible to build a utility function that could be patched in. 

It’s a limited implementation (Shell), it didn’t work for nested statements

**Reply by gswifort** [reply]
I understand, you want to replicate the tasks performed in [genDoc](https://github.com/CEXT-Dan/PyRx/blob/main/PyRxStubs/utility/genDoc.py)..? However, I see that you use json files there, how are they generated?

It's true that `PyShell` has some shortcomings, but unfortunately I haven't found anything better, especially since we seem to be limited to `wxpython`.

**Reply by CEXT-Dan** [reply]
> getCorner( (Point3d)arg1, (str)arg2) -> tuple :
>     ![(basePt: PyGe.Point3d,prompt: str)]!<[{10840}]>
>     C++ signature :
>         class boost::python::tuple getCorner(class AcGePoint3d,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >)
> 

example 
base function getCorner( (Point3d)arg1, (str)arg2) -> tuple :
argruments wrapped in tags  ![(basePt: PyGe.Point3d,prompt: str)]!
Then the JSON is just a map, where the ID is the key   {10840 : “do something”}

I replace the arguments from the auto generated signature, so it becomes 
getCorner(basePt: PyGe.Point3d,prompt: str) -> tuple :
   ‘’’ do something’’’

**Reply by gswifort** [reply]
So to make a patch for Shell I will have to have access to the json files and perform roughly the same steps as in `genDoc`, but instead of writing a `pyi` file I will have to call `Ed.Editor.getCorner.__doc__ = docstring`

**Reply by CEXT-Dan** [reply]
For now, seems like the easiest. 
It’s been a real battle from a stub perspective, mostly because of my incompetence

**Reply by gswifort** [reply]
Ok, so I'll work on it.

**Reply by gswifort** [reply]
Do you have any wishes regarding the behavior of the tooltip window?  I think there should be an option to turn it off (it can be annoying with small palette sizes).  Unless its position can be better controlled.  I also think that the previous line editing mode should be activated by the up arrow and not Ctrl + up arrow.

**Reply by CEXT-Dan** [reply]
Not really, though, there should be some  control over setting the tool tip position. 
If there’s an event, something like ontooltip, it would be possible to send messages to it, like move and size

**Comment by gswifort** [comment]
The pyshell palette loses focus when the tooltip is displayed and the cursor is on the tooltip. When the cursor is out, it's ok, but you can't see what you're writing.

**Reply by CEXT-Dan** [reply]
AutoCAD doesn't lose focus, but sometimes the command line is covered. something like this seems to work
```Python
import traceback

import wx
from pyrx_imp import *
from wx.py.shell import Shell

host = Ap.Application.hostAPI()
palette = Ap.PaletteSet("PyShell","{b98f3a34-ed66-4c3f-9c6a-9804e13b18b0}")

class MyPanel(wx.Panel):
    def __init__(self):
        super().__init__()
        self.Bind(wx.EVT_SHOW, self.OnShow)
        
    def OnShow(self, event: wx.ShowEvent):
        sizer = wx.BoxSizer()
        self.shellcrl = Shell(self, locals=globals())
        sizer.Add(self.shellcrl, 1, wx.EXPAND, 0)
        self.SetSizerAndFit(sizer)
        self.Layout()
        self.shellcrl.Bind(wx.EVT_CHAR, self.OnChar)
        
    def OnChar(self, event: wx.KeyEvent):
        try:
            if self.shellcrl.CallTipActive():
                self.shellcrl.CallTipCancel()
        finally:
            event.Skip()
            
def PyRxCmd_pyshell():
    try:
        panel = MyPanel()
        palette.add("PyShell", panel)
        palette.setVisible(True)
    except Exception:
        traceback.print_exc()
```

**Comment by gswifort** [comment]
> https://github.com/CEXT-Dan/PyRx/discussions/105#discussioncomment-10459000  

I installed [python 3.13](https://www.python.org/downloads/release/python-3130rc1/#:~:text=Windows%20installer%20(64%2Dbit))
the console is improved but still doesn't show signatures and documentation - `ptpython` seems better.

@CEXT-Dan,  how do you feel about using the python interpreter in the console instead of the palette?

**Reply by CEXT-Dan** [reply]
I’m not sure I follow, the python interpreter is embedded in the CAD process, palettes are part of that process.
If you redirect stdin,stdout, CAD will most likely fail. If you have something working, that would be cool

I’m not actually using the palette, I had changed the PyRx tests to output to log files, this allows me to use windiff across all the cad versions

**Reply by gswifort** [reply]
of course I have :)

```py
import sys
import traceback

import win32console
from pyrx_imp import Ap, Db, Ed, Ge, Gi, Gs, Rx


def PyRxCmd_doit():
    try:
        main()
        print("\nBack to CAD")
    except Exception:
        traceback.print_exc()


def main():
    try:
        stdout = sys.stdout
        stderr = sys.stderr
        stdin = sys.stdin
        win32console.AllocConsole()
        sys.stdout = sys.stderr = open("CONOUT$", "w")
        sys.stdin = open("CONIN$", "r")

        status, distance = Ed.Editor.getDist("getDist")
        print(f"{distance=}")
        input("Press any key to continue . . .")

    finally:
        sys.stdout = stdout
        sys.stderr = stderr
        sys.stdin = stdin
        win32console.FreeConsole()
```

**Reply by gswifort** [reply]
Eeverything works until I run `main` in a separate thread.

```raw
    status, distance = Ed.Editor.getDist("getDist")
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: unidentifiable C++ exception
```

**Reply by CEXT-Dan** [reply]
Yep, none of the CAD's are thread safe. There are probably ways to do this, none of them will be easy. IMHO, the console emulation window should be a child of the main frame, be able to hook into the host application’s message pump

**Reply by gswifort** [reply]
Ok, so I think it's enough to run the REPL in a new thread, but send the code to be executed to the main thread.

```py
import queue
import sys
import threading
import traceback
from code import InteractiveConsole

import win32api
import win32con
import win32console
import win32gui
from pyrx_imp import Ap


class Repl(InteractiveConsole):
    WND_PROC_MSG = 1111

    def __init__(self, locals=None, filename="<console>") -> None:
        super().__init__(locals, filename)
        self._res_queue = queue.Queue(1)
        self._code_queue = queue.Queue(1)

    def _runcode_in_main_thread(self):
        code = self._code_queue.get_nowait()
        try:
            res = super().runcode(code)
        except Exception as e:
            self._res_queue.put((False, e))
        else:
            self._res_queue.put((True, res))

    def runcode(self, code) -> None:
        self._code_queue.put(code)
        win32gui.PostMessage(Ap.Application.mainWnd(), self.WND_PROC_MSG, 0, 0)
        success, res = self._res_queue.get()
        if success:
            return res
        else:
            raise res

    def is_repl_msg(self, msg, w_param, l_param):
        return msg == self.WND_PROC_MSG

    def my_wnd_proc(self, hwnd, msg, w_param, l_param):
        if self.is_repl_msg(msg, w_param, l_param):
            self._runcode_in_main_thread()
            return 0
        return win32gui.CallWindowProc(self._old_wnd_proc, hwnd, msg, w_param, l_param)

    def _add_wnd_proc_hook(self):
        self._old_wnd_proc = win32gui.SetWindowLong(Ap.Application.mainWnd(), win32con.GWL_WNDPROC, self.my_wnd_proc)

    def _restore_wnd_proc_hook(self):
        win32api.SetWindowLong(Ap.Application.mainWnd(), win32con.GWL_WNDPROC, self._old_wnd_proc)

    def run(self, *args, **kwargs):
        self._add_wnd_proc_hook()
        try:
            self.interact(*args, **kwargs)
        finally:
            self._restore_wnd_proc_hook()


def run_in_thread():
    try:
        win32console.FreeConsole()
        win32console.AllocConsole()
        stderr = sys.stderr
        stdout = sys.stdout
        stdin = sys.stdin
        conin = open("CONIN$", "r")
        conout = open("CONOUT$", "w")
        sys.stdout = sys.stderr = conout
        sys.stdin = conin

        repl = Repl()
        repl.run()

    finally:
        sys.stderr = stderr
        sys.stdout = stdout
        sys.stdin = stdin
        win32console.FreeConsole()


def PyRxCmd_repl():
    try:
        t = threading.Thread(target=run_in_thread)
        t.start()
    except Exception:
        traceback.print_exc()
```

It seems that the `wx.CallAfter` function is used for the same purpose, but `wx.GetApp()` would have to return the same as `Ap.Application.wxApp()`

**Reply by gswifort** [reply]
@CEXT-Dan, are you planning to overwrite `wx.CallAfter`?

**Reply by CEXT-Dan** [reply]
I’m not familiar with that function, so I guess I have no plans.

**Reply by gswifort** [reply]
To call a function in the main thread I need to use a unique `msg` in `my_wnd_proc`, the system reserves the range to `WM_USER` (1024) does AutoCAD have a reserved range?

**Reply by CEXT-Dan** [reply]
There’s no set range for AutoCAD, i.e. WM_ACAD_KEEPFOCUS = 28929
There’s always the possibility of a conflict, here’s some ideas:
https://learn.microsoft.com/en-us/windows/win32/winmsg/wm-user

somewhere in WM_USER through 0x7FFF

**Comment by gswifort** [comment]
I am uploading a module to test the REPL. The original file is [here](https://github.com/gswifort/PyRx/blob/repl/pyrx/repl.py), but in the attached file I have added all the functions from other `pyrx` modules. The only additional dependency is `ptpython` (apart from `pywin32`). **NOTE**: after finishing work in the console it is necessary to call the `exit()` command, closing the console window will immediately close the CAD application.
Two REPLs are implemented:

- `STDREPL` - based on the REPL from the standard library
- `PTREPL` - based on `ptpython`

[repl.zip](https://github.com/user-attachments/files/16887620/repl.zip)

**Comment by CEXT-Dan** [comment]
Looks pretty cool! Though, for me I don’t see that ptpython is any better than the wxPython shell, in that it does not see the .PYI stubs

**Reply by gswifort** [reply]
That's right, unfortunately it doesn't read docstrings from pyi, but type hints do (you have to enable them in the settings - F2). For me, the advantage is that it copes better with syntax hinting (nesting) and can work with other console tools (e.g. output formatting and coloring). I will still work on overwriting docstrings.

**Comment by gswifort** [comment]
#237 
#240
